### PR TITLE
Support puppetjs reconnection events. puppetjs/puppetjs#71

### PR DESCRIPTION
**Pull request by joozek78**
cc @tomalec @warpech

### CODE REVIEW

**Comment by warpech** [comment]
@tomalec could you review this today?

File: puppet-client.html [hunk 1]
```diff
@@ -29,6 +29,49 @@
...
+            .error pre {
+                white-space: pre-wrap;
+                text-align: left;
+                padding: 0 15px;
+            }
+        </style>
+        <div class="box-container" hidden$="{{!showReconnectingIn}}"><div class="box notice"><strong>Not connected.</strong> Connecting in {{reconnectionSeconds}}s... <a href="" on-tap="reconnectNow">Retry now</a></div></div>
+        <div class="box-container" hidden$="{{!showReconnectingNow}}"><div class="box notice"><strong>Connecting now...</strong></div></div>
+        <div class="box-container" hidden$="{{!connectionError.length}}"><div class="box error"><strong>Connection error:</strong> <pre>{{connectionError}}</pre><a href="" on-tap="reload">Reload page</a> </div></div>
+    </template>
```
Comment (tomalec) [inline puppet-client.html]: Do we really want to keep those DOM notifications here? What about i18n, what  if some app would like to show them elsewhere with different messages, consistently with their messages?

I would say even we would like to show all Launcher related messages in the same place and manner.

File: puppet-client.html [hunk 2]
```diff
@@ -162,16 +199,46 @@
                 value: '/'
             },
 
+            /**
+             * @private
```
Comment (tomalec) [inline puppet-client.html]: What do you mean by private here?

File: puppet-client.html [hunk 3]
```diff
@@ -162,16 +199,46 @@
...
+             * show the "reconnecting in Xs..." box
+             */
+            showReconnectingIn: {
+                type: Boolean,
+                value: false
```
Comment (tomalec) [inline puppet-client.html]: What about `readOnly: true` here, and all properties bellow?

File: puppet-client.html [hunk 4]
```diff
@@ -29,6 +29,49 @@
...
+                background-color: #fad7d4;
+                border-color: #f08a81;
+            }
+            .error pre {
+                white-space: pre-wrap;
+                text-align: left;
+                padding: 0 15px;
+            }
+        </style>
+        <div class="box-container" hidden$="{{!showReconnectingIn}}"><div class="box notice"><strong>Not connected.</strong> Connecting in {{reconnectionSeconds}}s... <a href="" on-tap="reconnectNow">Retry now</a></div></div>
```
Comment (tomalec) [inline puppet-client.html]: WDYT about `mute` or `silent` attribute and wrapping it with `<template is="dom-if" if="{{!mute}}>`

File: puppet-client.html [hunk 5]
```diff
@@ -212,7 +279,18 @@
                     }
                 }.bind(this)
             });
+            var that = this;
             this.puppet.addEventListener('patch-applied', this.onPatchApplied.bind(this));
+            this.puppet.addEventListener('reconnection-countdown', function (ev) {
```
Comment (tomalec) [inline puppet-client.html]: Shouldn't we forward those events? Especially they were supposed to be bubbling

Comment (joozek78) [inline puppet-client.html]: Okay, you're probably right

File: puppet-client.html (hunk 1 above)
Comment (joozek78) [inline puppet-client.html]: What exactly do you propose? Up until now the message was hardcoded

File: puppet-client.html (hunk 2 above)
Comment (joozek78) [inline puppet-client.html]: this property is used internally and wasan't meant to be exposed. We (together with @warpech ) just didn't find a way not to expose it

File: puppet-client.html (hunk 3 above)
Comment (joozek78) [inline puppet-client.html]: that would do!

File: puppet-client.html (hunk 4 above)
Comment (joozek78) [inline puppet-client.html]: I think if we provide means to override this behavior (what you proposed with events) that would be enough

File: puppet-client.html (hunk 2 above)
Comment (tomalec) [inline puppet-client.html]: https://www.polymer-project.org/1.0/docs/devguide/properties
I think, `readOnly: true, notify: false` is worth checking.

**Comment by tomalec** [comment]
I think we should solve the problem of hard-coded notifications (text and style).

## Use cases:
1. Developer uses just `<puppet-client></puppet-client>` and would like to see the notifications
2. Developer would like to use `<puppet-client></puppet-client>` but with his/her own message with custom styles
3. Developer would like to use `<puppet-client></puppet-client>` but would like to integrate it with his/her own notification system.

---
My idea is to make it HTML-way :)
## Usage:
1. 

 ```html
<puppet-client></puppet-client>
```
:) and this should show the default message box

2. Developer provides the content with his/her own message styled and composed with CSS & HTML

 ```html
<puppet-client>
       <div slot="reconnectingNow" class="my-fancy-message">Chwileczkę, <span style="color: cyan;">łączę</span></div>
        ...
</puppet-client>
```
3. Developer provides the content with his/her own message styled and composed with CSS & HTML

 ```html
<puppet-client no-messages></puppet-client>
<script>
      document.querySelector('puppet-client').addEventListener('reconnection-countdown', doCustomMagic)
</script>
```

---
## Implementation

It should be easy to achieve with conditional template and fallback content for slots

```html
<slot name="reconnectingNow">
    <div class="box-container" hidden$="{{!showReconnectingIn}}"><div class="box notice"><strong>Not connected.</strong> Connecting in {{reconnectionSeconds}}s... <a href="" on-tap="reconnectNow">Retry now</a></div></div>
</slot>
```
(same should work for `<content>` tag)

**Comment by tomalec** [comment]
Or even otherwise.
Make `polymer-client` non-visual element like `<link is="puppet-client"></link>` 
As basically we would like to just link our document to data-model.

Then `<puppet-messages>` element will listen to the events and show the messages as described above in 1. and 2..

Thanks to that `link is="puppet-client"` could be use anywhere and will not be treat as something to display or layout (by `juicy-composition`), and `puppet-messages` could be place in more accurate place.

@warpech WDYT?

**Comment by joozek78** [comment]
Together with warpech we figured out that current behavior (event forwarding + checking ev.handled) is enough customization for the first version:

* Current launcher will display reconnection message and will override fatal connection error message with its own (please reload)
* If somebody wants to override connection messages he can do it in similar fashion

File: puppet-client.html (hunk 2 above)
Comment (joozek78) [inline puppet-client.html]: @tomalec but why use `notify: false`? it has nothing to do with read-access

Comment (tomalec) [inline puppet-client.html]: To reduce the overhead of publishing it to polymer notification protocol, as we do not expose and notify about each change. We already emit puppet events.

Comment (joozek78) [inline puppet-client.html]: Is it on by default? [docs](https://www.polymer-project.org/1.0/docs/devguide/data-binding#property-notification) doesn't state it one way or the other, but I though it's off

Comment (tomalec) [inline puppet-client.html]: I thought it was on by default, but it seems you're right 
>  (If the property being bound does not have the notify flag set, only one-way (downward) binding will occur.)

**Pull request merged**